### PR TITLE
Add PHP-Parser 5 support

### PR DESCRIPTION
**Pull request by sidz**
This PR:

- [x] Adds PHP-Parser 5 support. ([Upgrade guide](https://github.com/nikic/PHP-Parser/blob/master/UPGRADE-5.0.md))
- Fixes https://github.com/infection/infection/issues/1918

### CODE REVIEW

File: src/Mutator/Operator/Finally_.php [hunk 1]
```diff
@@ -95,6 +95,6 @@ private function hasAtLeastOneCatchBlock(Node $node): bool
         $parentNode = ParentConnector::getParent($node);
         Assert::isInstanceOf($parentNode, Node\Stmt\TryCatch::class);
 
-        return count($parentNode->catches) > 0;
+        return $parentNode->catches !== [];
```
Comment (theofidry) [inline src/Mutator/Operator/Finally_.php:98]: should this really be changed? I don't feel strongly about using this form rather than `count()` but IMO should be done on the whole codebase in a dedicated PR rather than here

File: src/Mutator/Removal/MatchArmRemoval.php [hunk 2]
```diff
@@ -116,13 +117,13 @@ public function mutate(Node $node): iterable
         foreach ($node->arms as $i => $arm) {
             $arms = $node->arms;
 
-            if ($arm->conds !== null && count((array) $arm->conds) > 1) {
+            if (count((array) $arm->conds) > 1) {
```
Comment (theofidry) [inline src/Mutator/Removal/MatchArmRemoval.php:120]: looks like an unrelated change?

File: src/Mutator/Removal/MatchArmRemoval.php [hunk 3]
```diff
@@ -116,13 +117,13 @@ public function mutate(Node $node): iterable
...
                 foreach ($arm->conds as $j => $cond) {
                     $conds = $arm->conds;
```
Comment (theofidry) [inline src/Mutator/Removal/MatchArmRemoval.php:122]: unrelated, so best in a different PR, but since the value is used L120 I think this assignment should precede L120, which can then make use of this variable

File: src/Mutation/Mutation.php [hunk 4]
```diff
@@ -142,10 +141,9 @@ public function getMutatedNode(): MutatedNode
         return $this->mutatedNode;
     }
 
-    // TODO: hasTest()?
     public function isCoveredByTest(): bool
     {
-        return $this->coveredByTests;
+        return $this->tests !== [];
```
Comment (theofidry) [inline src/Mutation/Mutation.php:146]: this looks like an unrelated change and the behaviour is different: previously this value was memoized, now it will be re-evaluated on every call

File: src/Mutation/Mutation.php [hunk 5]
```diff
@@ -162,12 +160,12 @@ public function getAllTests(): array
     public function getNominalTestExecutionTime(): float
     {
         // TestLocator returns non-unique tests, and JUnitTestCaseSorter works around that; we have to do that too.
-        return $this->nominalTimeToTest ?? $this->nominalTimeToTest = (new JUnitTestCaseTimeAdder($this->tests))->getTotalTestTime();
+        return $this->nominalTimeToTest ??= (new JUnitTestCaseTimeAdder($this->tests))->getTotalTestTime();
```
Comment (theofidry) [inline src/Mutation/Mutation.php:163]: I'm surprised Rector does not capture this

**Review by theofidry — COMMENTED** [review]
👍 for the PR. I think however there is a few unrelated changes that could be extracted for clarity.

I am also a bit confused, is the PHP-Parser API changes working on 4.x? In which case we could migrate to it before allowing PHP-Parser 5.x

**Comment by sidz** [comment]
> I am also a bit confused, is the PHP-Parser API changes working on 4.x?
The main issue in tests so we will have to support 2 sets of tests in some cases. Also it looks like we will have to override Pretty Printer

**Comment by ttomdewit** [comment]
When I try to install with this branch it also signals that it needs to update `sebastian/diff` to `^6.0.0`.

```
  Problem 2
    - Root composer.json requires infection/infection dev-add-php-parser-5-support@dev -> satisfiable by infection/infection[dev-add-php-parser-5-support].
    - infection/infection dev-add-php-parser-5-support requires sebastian/diff ^3.0.2 || ^4.0 || ^5.0 -> satisfiable by sebastian/diff[3.0.2, 3.0.3, 3.0.4, 4.0.0, ..., 4.0.5, 5.0.0, ..., 5.1.0].
    - You can only install one version of a package, so only one of these can be installed: sebastian/diff[3.0.2, 3.0.3, 3.0.4, 4.0.0, ..., 4.0.5, 5.0.0, ..., 5.1.0, 6.0.0].
    - phpunit/phpunit[11.0.0, ..., 11.0.2] require sebastian/diff ^6.0 -> satisfiable by sebastian/diff[6.0.0].
    - Root composer.json requires phpunit/phpunit ^11.0.0 -> satisfiable by phpunit/phpunit[11.0.0, 11.0.1, 11.0.2].

```

**Comment by sidz** [comment]
`sebastian/diff` released 7 days ago with phpunit 11. As an option you could downgrade phpunit.

also we have openned PR to add support for `sebastian/diff: ^6.0` https://github.com/infection/infection/pull/1913

**Comment by ttomdewit** [comment]
> `sebastian/diff` released 7 days ago with phpunit 11. As an option you could downgrade phpunit.
> 
> also we have openned PR to add support for `sebastian/diff: ^6.0` #1913

Sorry I completely missed that one! I was aiming to get PHPUnit 11 running in my project, hence me trying this branch. Best of luck wrapping up, looking forward to upgrading!

**Comment by darthf1** [comment]
Is this ready now that https://github.com/infection/infection/pull/1913 is merged? :)

**Comment by sidz** [comment]
> The main issue in tests: we will have to support 2 sets of tests in some cases. Also it looks like we will have to override Pretty Printer

What about dropping support for PHP-Parser 4 in release 0.30? @infection/core

**Comment by maks-rafalko** [comment]
I have to say that I'm not understanding the current issues with tests and pretty printer yet, but I'm completely ok with dropping support for php-parser v4. 

In the worst case if some project is stuck with php-parser v4 and want new Infection - infection.phar can be used.

**Comment by maks-rafalko** [comment]
@sidz please review :smiley:

**Review by maks-rafalko — APPROVED** [review]

**Review by sidz — COMMENTED** [review]
LGTM @maks-rafalko

**Comment by maks-rafalko** [comment]
Thanks for your hard work @sidz, let's merge and release

**Comment by maks-rafalko** [comment]
Released https://github.com/infection/infection/releases/tag/0.28.0
